File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    dic = {i: 1 for i in phone_book}
    # O(20*n) = O(n)
    for phone in phone_book:
        for i in range(len(phone)):
            if phone[:i] in dic:
                answer = False
    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

bool solution(vector<string> phone_book) {
    map<string, int> prefix;
    
    for (string p : phone_book)
    {
        for (int i = p.size(); i > 0; i--)
            prefix[p.substr(0, i)]++;
    }
    for (string p : phone_book)
    {
        if (prefix[p] != 1)
            return false;
    }
    return true;
}